syntax = "proto3";

package helloworld;

message ComplexType {
  int64 id = 1;
  string name = 2;
}

message ComplexType2 {
  int64 id = 1;
  int64 age = 2;
  string address = 3;
  string name = 4;
  string food = 5;
  string color = 6;
  string game = 7;
  string sport = 8;
  string height = 9;
  string nickname = 10;
}

// Define the Hello service
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}

  rpc VoidFunction (VoidRequest) returns (VoidReply) {}

  rpc LongFunction (LongOperationRequest) returns (LongOperationReply) {}

  rpc Long8Function (Long8OperationRequest) returns (LongOperationReply) {}

  rpc StringOperation (StringOperationRequest) returns (StringOperationReply) {}

  rpc ComplexOperation (ComplexOperationRequest) returns (ComplexOperationReply) {}

  rpc ComplexOperation2 (ComplexOperation2Request) returns (ComplexOperation2Reply) {}
  
  rpc MultipleString (MultipleStringRequest) returns (MultipleStringReply) {}

}

// The request message containing the user's name
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

message VoidRequest {}

message VoidReply {}

message LongOperationRequest {
  int64 arg1 = 1;
}

message Long8OperationRequest {
  int64 arg1 = 1;
  int64 arg2 = 2;
  int64 arg3 = 3;
  int64 arg4 = 4;
  int64 arg5 = 5;
  int64 arg6 = 6;
  int64 arg7 = 7;
  int64 arg8 = 8;
}

message LongOperationReply {
  int64 result = 1;
}

message StringOperationRequest {
  string input_string = 1;
}

message StringOperationReply {
  string output_string = 1;
}

message ComplexOperationRequest {
  ComplexType input_complex = 1;
}

message ComplexOperationReply {
  ComplexType output_complex = 1;
}

message ComplexOperation2Request {
  ComplexType2 input_complex = 1;
}

message ComplexOperation2Reply {
  ComplexType2 output_complex = 1;
}


message MultipleStringRequest{
  string s1 = 1;
  string s2 = 2;
  string s3 = 3;
  string s4 = 4;
  string s5 = 5;
  string s6 = 6;
}

message MultipleStringReply{
  string s1 = 1;
}